### PR TITLE
[ Misc ] Support Fp8 via `llm-compressor`

### PR DESCRIPTION
**Pull request by robertgshaw2-redhat**
SUMMARY:
* refactors fp8 utilities for shared use across `autofp8`, `compressed-tensors`, `nvidia-ammo`
* adds support for `compressed-tensors` Fp8 models (dense only)

Relevant PR: https://github.com/vllm-project/vllm/compare/main...pavanimajety:vllm:ammo_ckpt_rev2
* cc @pavanimajety this will help land nvidia-ammo

TEST PLAN:
* added regression test to `lm-eval-harness` automation

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by pavanimajety** [comment]
@robertgshaw2-neuralmagic Thanks for the cc and the utils you created. Here's the WIP PR: https://github.com/vllm-project/vllm/pull/6112

Do we have an agreement about adding the "quant_source" parameter in "quantization_config" for static scaled checkpoints?

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py [hunk 1]
```diff
@@ -118,6 +118,40 @@ def _is_dynamic_token_w8a8(self, weight_quant: BaseModel,
...
+                     input_quant: BaseModel) -> bool:
+        # Confirm weights and activations quantized.
+        if weight_quant is None or input_quant is None:
+            return False
+
+        # Confirm we have floating points.
+        if not (weight_quant.type == QuantizationType.FLOAT
+                and input_quant.type == QuantizationType.FLOAT):
+            return False
+
```
Comment (dsikka) [inline vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py:131]: Could we clean this up a bit so it's easier to follow what is and what isn't supported?

Maybe starting with something like this for the lines above?
```
is_weight_quant_float = weight_quant and weight_quant.type == `QuantizationType.FLOAT`
is_input_quant_float = input_quant and input_quant.type == `QuantizationType.FLOAT`
```

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py [hunk 2]
```diff
@@ -118,6 +118,40 @@ def _is_dynamic_token_w8a8(self, weight_quant: BaseModel,
...
+        # Confirm activation scheme is supported.
+        is_symmetric_activation = input_quant.symmetric
+        is_per_tensor_activation = (
+            input_quant.strategy == QuantizationStrategy.TENSOR)
+        if not (is_symmetric_activation and is_per_tensor_activation):
+            return False
+
+        # All conditions satisfied.
+        return True
```
Comment (dsikka) [inline vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py:153]: Could be combine the above `is_xxx` variables and make sure all True and return false otherwise?

Slightly easier to follow

**Review by dsikka — CHANGES_REQUESTED** [review]

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py [hunk 3]
```diff
@@ -148,7 +182,12 @@ def _get_schema(self, weight_quant: BaseModel,
                     strategy=weight_quant.strategy,
                     group_size=weight_quant.group_size)
 
-        if self.quant_format == CompressionFormat.int_quantized.value:
+        if (self.quant_format == CompressionFormat.int_quantized.value or
+                self.quant_format == CompressionFormat.float_quantized.value):
+            if self._is_fp8_w8a8(weight_quant, input_quant):
```
[the remark this answers is not in the document]
Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py:187]: I dont understand why we have `CompressionFormat`

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py (hunk 1 above)
Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py:131]: yep

**Comment by robertgshaw2-neuralmagic** [comment]
> @robertgshaw2-neuralmagic Thanks for the cc and the utils you created. Here's the WIP PR: #6112
> 
> Do we have an agreement about adding the "quant_source" parameter in "quantization_config" for static scaled checkpoints?

@pavanimajety - can you share a sample model on the HF hub of what it would look like?

Let me get this merged in + then we can use these utils for `ModelOpt`

File: vllm/model_executor/layers/quantization/utils/fp8_utils.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    capability = capability[0] * 10 + capability[1]
+
+    return ops.cutlass_scaled_mm_supports_fp8(capability)
+
+
+def per_tensor_quantize(tensor: torch.Tensor,
+                        inv_scale: Union[float, torch.Tensor]) -> torch.Tensor:
+    finfo = torch.finfo(torch.float8_e4m3fn)
+    qweight = (tensor / inv_scale).clamp(min=finfo.min, max=finfo.max)
+    return qweight.to(torch.float8_e4m3fn)
```
Comment (comaniac) [inline vllm/model_executor/layers/quantization/utils/fp8_utils.py:21]: Actually we should be able to use `ops.scaled_fp8_quant` to replace this function?

File: vllm/model_executor/layers/quantization/utils/fp8_utils.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+            end = start + logical_width
+            weight_dq = per_tensor_dequantize(weight[start:end, :],
+                                              weight_scale[idx])
+            weight[start:end, :] = per_tensor_quantize(weight_dq, max_w_scale)
+            start = end
+
+    return max_w_scale, weight
+
+
+def fp8_apply(
```
Comment (comaniac) [inline vllm/model_executor/layers/quantization/utils/fp8_utils.py:78]: Ditto. Better to be something like `apply_fp8_linear`

**Comment by pavanimajety** [comment]
> Let me get this merged in + then we can use these utils for `ModelOpt`

We are little delayed on the model available on HF HUB. But here's the quantization config for us: 

```
 "quantization_config": {
     "activation_scheme": "static",
     "quant_method": "fp8",
     "quant_source":"modelopt"
 }
```

File: vllm/model_executor/layers/quantization/fp8.py [hunk 6]
```diff
@@ -11,22 +11,17 @@
 from vllm.model_executor.layers.linear import LinearBase, LinearMethodBase
 from vllm.model_executor.layers.quantization.base_config import (
     QuantizationConfig, QuantizeMethodBase)
+from vllm.model_executor.layers.quantization.utils.fp8_utils import (
+    all_close_1d, create_scale_param, cutlass_fp8_supported, fp8_apply,
+    per_tensor_dequantize, per_tensor_quantize, requantize_with_max_scale)
```
Comment (pavanimajety) [inline vllm/model_executor/layers/quantization/fp8.py:16]: nit: probably don't need the dequantize and quantize imports anymore.

**Review by pavanimajety — APPROVED** [review]

**Comment by robertgshaw2-neuralmagic** [comment]
> > Let me get this merged in + then we can use these utils for `ModelOpt`
> 
> We are little delayed on the model available on HF HUB. But here's the quantization config for us:
> 
> ```
>  "quantization_config": {
>      "activation_scheme": "static",
>      "quant_method": "fp8",
>      "quant_source":"modelopt"
>  }
> ```

@pavanimajety I would suggest `"quant_method": "modelopt"` in the config. This will allows us to know to use the `ModelOptLinearMethod` rather than going down the AutoFP8 pathway

**Comment by robertgshaw2-neuralmagic** [comment]
@comaniac @simon-mo this should be ready to go pending CI

File: vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_fp8.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+__all__ = ["CompressedTensorsW8A8Fp8"]
+
+
+class CompressedTensorsW8A8Fp8(CompressedTensorsScheme):
+
+    def __init__(self, input_dynamic: bool):
+        self.input_dynamic = input_dynamic
+        self.cutlass_fp8_supported = cutlass_fp8_supported()
+
+    # W8A8-Fp8 kernels support only per-tensor and per-channel cases.
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_fp8.py:21]: Just to be clear, in the per channel case there is no max reduction done? It isn't clear from the code in this function. Is per channel actually supported?

**Review by mgoin — APPROVED** [review]
Great job with the refactoring cleanup into utility functions here. This is a definite improvement now and sets a better structure for future variations, no qualms

File: vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_fp8.py (hunk 7 above)
Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_fp8.py:21]: The cutlass kernels do actually support the `per-channel` case. We do not allow this at the Python layer so far

File: vllm/model_executor/layers/quantization/utils/w8a8_utils.py [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+                                   **extra_weight_attrs) -> Parameter:
+    scale = Parameter(torch.empty((sum(output_partition_sizes), 1),
+                                  dtype=torch.float32),
+                      requires_grad=False)
+    scale[:] = torch.finfo(torch.float32).min
+    set_weight_attrs(scale, {"output_dim": 0, **extra_weight_attrs})
+    return scale
+
+
+def convert_to_channelwise(
```
Comment (comaniac) [inline vllm/model_executor/layers/quantization/utils/w8a8_utils.py:56]:
```suggestion
def convert_to_channelwise_scale(
```

**Review by comaniac — APPROVED** [review]
LGTM
